witchcraft/internal/tcpjson/tcp_logger.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpjson
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net"
+	"sync"
+
+	werror "github.com/palantir/witchcraft-go-error"
+)
+
+var _ io.Writer = (*TCPWriter)(nil)
+
+const ErrWriterClosed = "writer is closed"
+
+// TCPWriter writes logs to a TCP socket and wraps them with envelope metadata.
+type TCPWriter struct {
+	metadata LogEnvelopeMetadata
+	provider ConnProvider
+
+	// closedChan is used to signal that the writer is shutting down
+	closedChan chan struct{}
+
+	mu   sync.RWMutex // guards conn below
+	conn net.Conn
+}
+
+func NewTCPWriter(metadata LogEnvelopeMetadata, provider ConnProvider) *TCPWriter {
+	return &TCPWriter{
+		metadata:   metadata,
+		provider:   provider,
+		closedChan: make(chan struct{}),
+		conn:       nil,
+	}
+}
+
+func (d *TCPWriter) Write(logPayload []byte) (int, error) {
```

Comment:
I am starting to realize that this breaks the contract for an `io.Writer` since we end up writing more than `len(logPayload)` bytes to the underlying connection.

Reply:
I think this should be OK from a contract standpoint -- the API reports the number of input bytes written/processed, so the amount that we write to the underlying connection doesn't really matter.

However, it is true that the return value should reflect that (so in this example, if we write all the input, then it needs to return `len(logPayload)` rather than the number of byte that were written to the connection)

Reply:
For reference, https://stackoverflow.com/questions/32571504/what-number-of-bytes-written-should-my-io-writer-return-when-it-intentionally#comment52998337_32571504

Reply:
Thanks for clarifying this. I was getting tripped up thinking all io.Writer's had to return the total bytes actually written rather than the bytes processed of the original input.